### PR TITLE
array_position initial commit

### PR DESCRIPTION
**Pull request by kumarshail88**
This pull request adds a new scalar function to search the position of an element in array. All data types (including null) are possible to search that are supported by arrays as in Crate DB documentation.

Appropriate tests have been added and boundary conditions have been taken care of.

### CODE REVIEW

**Comment by seut** [comment]
ok to test

File: docs/general/builtins/scalar-functions.rst [hunk 1]
```diff
@@ -2718,6 +2718,58 @@ function supports arrays of any of the :ref:`primitive types
...
+.. _scalar-array_position:
+
+``array_position(anycompatiblearray, anycompatible [, integer ] ) → integer``
+-----------------------------------------------------------------------------
+
+The ``array_position`` function returns the subscript of the first
+occurrence of the second argument in the ``array``, or ``NULL`` if it's not
+present. If the third argument is given, the search begins at that subscript.
+The ``array`` must be one-dimensional. Comparisons are done using
+``IS NOT DISTINCT FROM`` semantics, so it is possible to search for ``NULL``.
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:2730]: CrateDB currently doesn't support the ``IS NOT DISTINCT FROM`` comparison statement. Thus we should phrase this differently to express than NULL values are supported.
Maybe something like:
"The function supports comparison of ``NULL`` values."

File: docs/general/builtins/scalar-functions.rst [hunk 2]
```diff
@@ -2718,6 +2718,58 @@ function supports arrays of any of the :ref:`primitive types
...
+
+::
+
+    cr> CREATE TABLE metrics(
+            ts timestamp,
+            values_sensor1 ARRAY(INTEGER),
+            values_sensor2 ARRAY(INTEGER)
+        );
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:2738]: Looks like the `ts` column is not needed for the examples, so I suggest to remove (keep scope low). Wouldn't also one array column be enough for all examples?

File: docs/general/builtins/scalar-functions.rst [hunk 3]
```diff
@@ -2718,6 +2718,58 @@ function supports arrays of any of the :ref:`primitive types
...
+    --Find out if a value is contained in an array
+
+    cr> SELECT array_position(values_sensor1, 10) IS NULL AS is_null
+        FROM metrics;
+    +---------+
+    | is_null |
+    +---------+
+    | TRUE    |
+    +---------+
+    SELECT 1 row in set (... sec)
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:2757]: As commented here https://github.com/crate/crate/issues/11943#issuecomment-983549996, using `ANY` is much more efficient for existence checks of array elements. Thus we should either include a similar comment to explain that `ANY` is a more performant alternative, or we may just remove this example here.

File: docs/general/builtins/scalar-functions.rst [hunk 4]
```diff
@@ -2718,6 +2718,58 @@ function supports arrays of any of the :ref:`primitive types
...
+
+::
+
+    --Find the position of an existing element.
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:2761]: Please move this out of the code block.

**Review by seut — COMMENTED** [review]
@kumarshail88 Thank you a lot for this contribution.
It looks great in general, I've added some detailed comments.

The documentation tests (doctests) are failing, see https://github.com/crate/crate/blob/master/devs/docs/tests.rst on how to run them locally.

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+    public static void register(ScalarFunctionModule scalarFunctionModule) {
+        scalarFunctionModule.register(
+            Signature.scalar(NAME,
+                parseTypeSignature("array(T)"),
+                parseTypeSignature("T"),
+                DataTypes.INTEGER.getTypeSignature()
+            )
+                .withTypeVariableConstraints(typeVariable("T"))
+                .withForbiddenCoercion()
```
[the remark this answers is not in the document]
Comment (kumarshail88) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java:40]: Thanks a lot @seut for review comments. I'll fix those asap. The reason behind forbidden coercion is to prevent searching of heterogenous types in the array. I thought, searching for text in an integer array wouldn't makes sense so the search parameter type should be same as inner type of the array. This also simply saves me from putting validation logic for the types inside the evaluate method business logic. 

Please help me understand the requirement here. If there is some other way to do it then yes we don't need coercion forbidden as well as those many signatures.

File: docs/general/builtins/scalar-functions.rst [hunk 6]
```diff
@@ -2718,6 +2718,58 @@ function supports arrays of any of the :ref:`primitive types
...
 
+.. _scalar-array_position:
+
+``array_position(anycompatiblearray, anycompatible [, integer ] ) → integer``
+-----------------------------------------------------------------------------
+
+The ``array_position`` function returns the subscript of the first
+occurrence of the second argument in the ``array``, or ``NULL`` if it's not
+present. If the third argument is given, the search begins at that subscript.
+The ``array`` must be one-dimensional. Comparisons are done using
```
[the remark this answers is not in the document]
Comment (kumarshail88) [reply docs/general/builtins/scalar-functions.rst:2729]: I think, it's working with multidimensional arrays too although somewhere in documentation I read that array(array(text)) is not supported so didn't add in documentation. I'll add it to doc.

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java:40]: When using a type variable constraint inside the signature, like you've defined by `.withTypeVariableConstraints(typeVariable("T"))`, the signature matching logic ensures that the types must match. 
With `.withForbiddenCoercion()` any implicit cast is prevented (CrateDB uses type precedence for that), which may useful for example when different function implementations for different types are needed/wanted.

File: docs/general/builtins/scalar-functions.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (seut) [reply docs/general/builtins/scalar-functions.rst:2729]: Yes right `array(array(text))` is not supported on various cases, e.g. our storage engine doesn't support that. But AFAIK these limitations aren't related to the `array_position` scalar, thus I wouldn't document it here.

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kumarshail88) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java:40]: Ok. Yes I did notice that implicit casting but for this function I choose to prevent that. So just to understand from few example. Let's say we have these calls
array_position([3,1,2,4,5,6] :: array(Integer) , 2.0) -> should return 3 as the target value is effectively equal? 

and what about this one
array_position([3,1,2,4,5,6] :: array(Integer) , '2') -> should return 3 as the target value is effectively equal? 

Although, I'll look deeper into it and bring some leniency to the function implementation w.r.t implicit cast.

File: docs/general/builtins/scalar-functions.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (kumarshail88) [reply docs/general/builtins/scalar-functions.rst:2729]: Sure. Will skip that. Thank you for answering too many questions. :)

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (seut) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java:40]: > array_position([3,1,2,4,5,6] :: array(Integer) , 2.0) -> should return 3 as the target value is effectively equal?

Yes. Due to type precedence rules, the array will be implicit cast to `array(double)` in that case.

> array_position([3,1,2,4,5,6] :: array(Integer) , '2') -> should return 3 as the target value is effectively equal?

Yes. All these cases work as long as implicit cast rules are sufficient for it. In that case, the string literal must be castable to an integer, providing e.g. 'foo' will fail with a cast error.

[the remark this answers is not in the document]
Comment (kumarshail88) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java]: Forbidden coercion has been removed.

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+public class ArrayPositionFunction<T> extends Scalar<Integer, List<T>> {
+
+    public static final String NAME = "array_position";
+    private final DataType returnType;
```
[the remark this answers is not in the document]
Comment (kumarshail88) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java]: Removed unused field.

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+import io.crate.data.Input;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+
+public class ArrayPositionFunction<T> extends Scalar<Integer, List<T>> {
```
[the remark this answers is not in the document]
Comment (kumarshail88) [reply server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java]: Removed generic type from class defn and using List<Object> as suggested.

File: server/src/test/java/io/crate/expression/scalar/ArrayPositionFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            , "Unknown function: array_position(_array('41', '98', 'hi', 'true', '@'), '@', 1.4), no overload found for matching argument types: (text_array, text, double precision).");
+    }
+
+    @Test
+    public void test_array_position_find_element_when_begin_position_given_wrong_type_text_throws_error() {
+        assertThrowsMatches(() -> assertEvaluate("array_position(['41','98', 'hi', 'true', '@'], '@', 'ab')", null),
+            UnsupportedOperationException.class
+            , "Unknown function: array_position(_array('41', '98', 'hi', 'true', '@'), '@', 'ab'), no overload found for matching argument types: (text_array, text, text).");
+    }
+
```
[the remark this answers is not in the document]
Comment (kumarshail88) [reply server/src/test/java/io/crate/expression/scalar/ArrayPositionFunctionTest.java:63]: Removed excess tests. Updated the tests as per review comments.

File: docs/general/builtins/scalar-functions.rst (hunk 1 above)
Comment (kumarshail88) [reply docs/general/builtins/scalar-functions.rst]: Updated in documents about null support.

File: docs/general/builtins/scalar-functions.rst (hunk 2 above)
Comment (kumarshail88) [reply docs/general/builtins/scalar-functions.rst]: I have simplified the function documentation. Removed verbose details taking inspiration from other functions. Please take a look again.

File: docs/general/builtins/scalar-functions.rst (hunk 3 above)
Comment (kumarshail88) [reply docs/general/builtins/scalar-functions.rst]: Added the ``TIP`` about the same.

File: docs/general/builtins/scalar-functions.rst (hunk 4 above)
Comment (kumarshail88) [reply docs/general/builtins/scalar-functions.rst]: Comment removed.

**Comment by kumarshail88** [comment]
Other than review comment changes,  I have also update the code to handle index out of bounds errors gracefully. If the third argument is out of range of the array then rather than throwing IndexOutOfBounds exception, function will simply ignore that optional field and will start the search from first subscript. Please let me know if that is okay.

**Comment by kumarshail88** [comment]
It took me a while to understand rebasing and squashing commits. Hope they are in order. :)

File: server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    private int getBeginPosition(Object position, int elementsSize) {
+
+        if (position == null) {
+            return 0;
+        }
+
+        int beginPosition = (int) position;
+        if (beginPosition < 1 || beginPosition > elementsSize) {
```
Comment (seut) [inline server/src/main/java/io/crate/expression/scalar/ArrayPositionFunction.java:93]: PostgreSQL returns `null` if the beginPosition is larger than the array size. This behaviour sounds logical to me as no value exists at these (larger) positions. I'd suggest to implement the same.

File: docs/general/builtins/scalar-functions.rst [hunk 11]
```diff
@@ -2748,6 +2748,45 @@ function supports arrays of any of the :ref:`primitive types
...
+    | position |
+    +----------+
+    |        3 |
+    +----------+
+    SELECT 1 row in set (... sec)
+
+.. TIP::
+    Using the :ref:`ANY <sql_any_array_comparison>` operator to search the
+    existence of the ``array`` elements is much more efficient and
+    recommended over using the ``array_position`` function.
```
Comment (seut) [inline docs/general/builtins/scalar-functions.rst:2787]:
```suggestion
    When searching for the existence of an ``array`` element, using the 
    :ref:`ANY <sql_any_array_comparison>` operator inside the ``WHERE``
    clause is much more efficient as it can utilize the index whereas  
    ``array_position`` won't even when used inside the ``WHERE`` clause.
```

**Review by seut — COMMENTED** [review]
Sorry for the late reply, somehow this slipped through.
I've added a final comment related to the behaviour and a more minor documentation suggestion.

File: docs/general/builtins/scalar-functions.rst (hunk 11 above)
Comment (mkleen) [reply docs/general/builtins/scalar-functions.rst]: Already addressed.

**Comment by mkleen** [comment]
This looks to me that everything is already addressed and ready to merge.
